File: src/engine/adapters/clickhouse/mutate.ts
import type { MutationPlan, MutationResult, MutationRowOp } from '../../../shared/domain/mutations';
import { encodePath } from '../../../shared/domain/tree';
import type { OpCtx } from '../adapter';
import { AdapterError, assertWritable } from '../errors';
import { assertColumnsKnown } from '../sql-mutate';
import * as catalog from './catalog';
import type { ClickHouseHandle } from './client';
import { runCatalogQuery, runCommand, type TrackQuery } from './query';
import { quoteIdent } from './read';

// D24/D25: MergeTree's PRIMARY KEY is a sparse index, not a unique key (F16) — there is no way to
// address "this one row" for an UPDATE or DELETE, so canUpdate/canDelete are permanently false
// (caps.ts) and every non-insert op is refused, mirroring kafka/produce.ts's assertInsert for the
// identical structural reason (an immutable log has no addressable row either).
function assertInsertOnly(
  op: MutationRowOp,
): asserts op is Extract<MutationRowOp, { kind: 'insert' }> {
  if (op.kind !== 'insert') {
    throw new AdapterError(
      'E_UNSUPPORTED',
      'ClickHouse only supports adding new rows (insert): a MergeTree PRIMARY KEY is a sparse ' +
        'index, not a unique key, so there is no addressable row to update or delete',
    );
  }
}

function resolveTablePath(path: MutationPlan['path']): { schema: string; table: string } {
  const [databaseSegment, objectSegment] = path.segments;
  if (
    path.segments.length !== 2 ||
    databaseSegment?.kind !== 'database' ||
    !objectSegment ||
    objectSegment.kind !== 'table'
  ) {
    throw new AdapterError(
      'E_NOT_FOUND',
      `mutate requires a database/table path, got: ${encodePath(path.segments)}`,
    );
  }
  return { schema: databaseSegment.name, table: objectSegment.name };
}

// D6/F27: backslash-then-quote — ClickHouse string literals use backslash escapes (F27), not
// SQL's standard doubled-quote convention every other adapter in this codebase renders with.
function literalFor(value: string | null): string {
  if (value === null) return 'NULL';
  return `'${value.replace(/\\/g, '\\\\').replace(/'/g, "\\'")}'`;
}

// D24: every insert op in one plan renders as a single multi-row INSERT statement — cheaper than
// one round trip per row (F19: ClickHouse is built for batch writes, not row-at-a-time OLTP), and
// what mutate() actually executes below, so preview() must render the same shape it would run
// (§8.13's "exact-command" contract) rather than one line per op. Columns are the UNION across
// every op, not assumed uniform: a plan's ops could in principle carry different value sets, and
// an op missing a given column pads that row's tuple with NULL rather than silently misaligning.
function renderInsert(
  relationSql: string,
  ops: Extract<MutationRowOp, { kind: 'insert' }>[],
  render: (value: string | null) => string,
): string {
  const columns: string[] = [];
  const seen = new Set<string>();
  for (const op of ops) {
    for (const col of Object.keys(op.values)) {
      if (!seen.has(col)) {
        seen.add(col);
        columns.push(col);
      }
    }
  }
  const columnSql = columns.map((c) => quoteIdent(c)).join(', ');
  const rows = ops.map((op) => {
    const tuple = columns.map((c) => render(op.values[c] ?? null)).join(', ');
    return `(${tuple})`;
  });
  return `INSERT INTO ${relationSql} (${columnSql}) VALUES ${rows.join(', ')}`;
}

// Synchronous (Adapter rule 3/§8.13): no catalog lookup, no network — trusts the plan's column
// names as given, exactly like every other adapter's own preview().
export function preview(plan: MutationPlan): string[] {
  const { schema, table } = resolveTablePath(plan.path);
  const relationSql = `${quoteIdent(schema)}.${quoteIdent(table)}`;
  plan.ops.forEach(assertInsertOnly);
  const inserts = plan.ops as Extract<MutationRowOp, { kind: 'insert' }>[];
  if (inserts.length === 0) return [];
  return [renderInsert(relationSql, inserts, literalFor)];
}

export async function mutate(
  h: ClickHouseHandle,
  ctx: OpCtx,
  track: TrackQuery,
  readOnly: boolean,
  plan: MutationPlan,
  nextQueryId: () => string,
): Promise<MutationResult> {
  // §8.12's standard: enforced here, not only greyed out in the UI (P5 D11).
  assertWritable(readOnly);

  const { schema, table } = resolveTablePath(plan.path);
  plan.ops.forEach(assertInsertOnly);
  const inserts = plan.ops as Extract<MutationRowOp, { kind: 'insert' }>[];
  if (inserts.length === 0) return { affectedRows: 0 };

  // Fresh in this same op (D7, mirrors resolveProjection's own discipline) — never trusts a
  // column name the renderer sent without re-checking it against the catalog right now.
  const exec: catalog.QueryExecutor = (sql, params) =>
    runCatalogQuery(h, ctx, sql, { queryId: nextQueryId() }, track, params);
  const target = await catalog.getReadTarget(exec, schema, table);
  for (const op of inserts) assertColumnsKnown(target.columns, Object.keys(op.values));

  const relationSql = `${quoteIdent(schema)}.${quoteIdent(table)}`;
  const sql = renderInsert(relationSql, inserts, literalFor);
  ctx.setCommand(sql);
  const { writtenRows } = await runCommand(h, ctx, sql, { queryId: nextQueryId() }, track);
  return { affectedRows: writtenRows > 0 ? writtenRows : inserts.length };
}

File: apps/kira-studio/internal/adapters/clickhouse/mutate.go
package clickhouse

import (
	"context"
	"strings"

	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/storage/model"
)

// assertInsertOnly is mutate.ts's own — D24/D25: MergeTree's PRIMARY KEY is a sparse index, not a
// unique key (F16) — there is no way to address "this one row" for an UPDATE or DELETE, so
// canUpdate/canDelete are permanently false (caps.go) and every non-insert op is refused.
func assertInsertOnly(op model.MutationRowOp) error {
	if op.Kind != "insert" {
		return adapters.New(adapters.CodeUnsupported,
			"ClickHouse only supports adding new rows (insert): a MergeTree PRIMARY KEY is a sparse "+
				"index, not a unique key, so there is no addressable row to update or delete", nil)
	}
	return nil
}

// literalFor is mutate.ts's own — D6/F27: backslash-then-quote, ClickHouse's own string-literal
// escaping (not SQL's standard doubled-quote convention every other adapter in this codebase
// renders with).
func literalFor(value *string) string {
	if value == nil {
		return "NULL"
	}
	escaped := strings.ReplaceAll(*value, `\`, `\\`)
	escaped = strings.ReplaceAll(escaped, `'`, `\'`)
	return "'" + escaped + "'"
}

// renderInsert is mutate.ts's own — D24: every insert op in one plan renders as a single multi-row
// INSERT statement, cheaper than one round trip per row, and what mutate() actually executes, so
// preview() must render the same shape it would run. Columns are the union across every op, not
// assumed uniform: an op missing a given column pads that row's tuple with NULL rather than
// silently misaligning.
func renderInsert(relationSQL string, ops []model.MutationRowOp, render func(*string) string) string {
	var columns []string
	seen := make(map[string]bool)
	for _, op := range ops {
		for _, kv := range op.Values {
			if !seen[kv.Name] {
				seen[kv.Name] = true
				columns = append(columns, kv.Name)
			}
		}
	}
	quotedColumns := make([]string, len(columns))
	for i, c := range columns {
		quotedColumns[i] = quoteIdent(c)
	}
	rows := make([]string, len(ops))
	for i, op := range ops {
		byName := make(map[string]*string, len(op.Values))
		for _, kv := range op.Values {
			byName[kv.Name] = kv.Value
		}
		tuple := make([]string, len(columns))
		for j, c := range columns {
			tuple[j] = render(byName[c])
		}
		rows[i] = "(" + strings.Join(tuple, ", ") + ")"
	}
	return "INSERT INTO " + relationSQL + " (" + strings.Join(quotedColumns, ", ") + ") VALUES " + strings.Join(rows, ", ")
}

// preview is mutate.ts's own — synchronous (Adapter rule 3/§8.13): no catalog lookup, no network.
func preview(plan model.MutationPlan) ([]string, error) {
	database, table, err := adapters.ResolveDatabaseTablePath(plan.Path)
	if err != nil {
		return nil, err
	}
	relationSQL := quoteIdent(database) + "." + quoteIdent(table)
	for _, op := range plan.Ops {
		if err := assertInsertOnly(op); err != nil {
			return nil, err
		}
	}
	if len(plan.Ops) == 0 {
		return []string{}, nil
	}
	return []string{renderInsert(relationSQL, plan.Ops, literalFor)}, nil
}

// mutate is mutate.ts's own.
func mutate(ctx context.Context, h *Handle, queryID string, op *adapters.OpCtx, track TrackQuery, readOnly bool, plan model.MutationPlan) (model.MutationResult, error) {
	if err := adapters.AssertWritable(readOnly); err != nil {
		return model.MutationResult{}, err
	}

	database, table, err := adapters.ResolveDatabaseTablePath(plan.Path)
	if err != nil {
		return model.MutationResult{}, err
	}
	for _, rowOp := range plan.Ops {
		if err := assertInsertOnly(rowOp); err != nil {
			return model.MutationResult{}, err
		}
	}
	if len(plan.Ops) == 0 {
		return model.MutationResult{AffectedRows: 0}, nil
	}

	// Fresh in this same op (D7, mirrors resolveProjection's own discipline) — never trusts a
	// column name the renderer sent without re-checking it against the catalog right now.
	target, err := getReadTarget(ctx, h, queryID, op, track, database, table)
	if err != nil {
		return model.MutationResult{}, err
	}
	for _, rowOp := range plan.Ops {
		if err := adapters.AssertColumnsKnown(target.Columns, rowOp.Values.Names()); err != nil {
			return model.MutationResult{}, err
		}
	}

	relationSQL := quoteIdent(database) + "." + quoteIdent(table)
	sql := renderInsert(relationSQL, plan.Ops, literalFor)
	op.SetCommand(sql)
	written, err := RunCommand(ctx, h, queryID, sql, op, track)
	if err != nil {
		return model.MutationResult{}, err
	}
	if written > 0 {
		return model.MutationResult{AffectedRows: int(written)}, nil
	}
	return model.MutationResult{AffectedRows: len(plan.Ops)}, nil
}
